llm/models/llm_provider.py

```
@@ -192,50 +196,75 @@ def action_fetch_models(self):
+
+                if not name:
+                    continue
+
+                # Skip duplicates
+                if name in wizard_models:
+                    continue
+                wizard_models.add(name)
+
+                # Determine model use and capabilities
+                capabilities = details.get("capabilities", ["chat"])
+                model_use = self._determine_model_use(name, capabilities)
+
+                # Check against existing models
+                existing = existing_models.get(name)
+                status = "new"
+                if existing:
+                    status = "modified" if existing.details != details else "existing"
+
+                lines_to_create.append(
+                    {
+                        "wizard_id": wizard.id,
+                        "name": name,
+                        "model_use": model_use,
+                        "status": status,
+                        "details": details,
+                        "existing_model_id": existing.id if existing else False,
+                        "selected": status in ["new", "modified"],
+                    }
+                )
+
+        except Exception as e:
+            # Check for authentication errors (incorrect API key)
+            error_msg = str(e)
+            if any(keyword in error_msg.lower() for keyword in [
+                'incorrect api key',
+                'invalid_api_key',
+                'authentication',
+                '401',
+                'unauthorized',
+                'invalid api key'
+            ]):
+                raise UserError(_(
+                    "Authentication failed. The provided API key appears to be incorrect. "
+                    "Please verify your API key and try again.\n\n"
+                    "Error details: %s"
+                ) % error_msg.split('\n')[0])  # Show only the first line for brevity
+            # Re-raise other exceptions with a user-friendly message
+            raise UserError(_(
```

We should not probably mask all exceptions under one UserError.. it would be better if we let them propagate and handle provider specific errors on provider level rather than here.

---

@adar2378 if `llm_ollama` is the only module that breaks, then maybe we could detect if this provider is ollama or not ?

the reason why i wanted to throw a UserError is because not everyone has the reflex to read entirely a traceback so not really knowing what's the issue (some may think that the module just does not work because a **traceback** is a too generic error..)

---

@moctarjallo I think you are right about user experience. But still, it should be handled on the provider-specific part, like in `llm_openai` or `llm_ollama` etc, modules and not here. Because it can hide bugs under the carpet of UserError.

For example, we could do something like this in llm_openai:
```python
# In llm_openai provider
  from openai import AuthenticationError, APIConnectionError, RateLimitError

  def openai_models(self, model_id=None):
      try:
          # ... fetch models
      except AuthenticationError as e:
          raise UserError(_("Invalid API key. Please check your OpenAI API key 
  configuration.\n\nDetails: %s") % e)
      except APIConnectionError as e:
          raise UserError(_("Could not connect to OpenAI API. Please check your network 
  connection.\n\nDetails: %s") % e)
      except RateLimitError as e:
          raise UserError(_("OpenAI rate limit exceeded. Please try again 
  later.\n\nDetails: %s") % e)
```

---

@adar2378 thanks for your feedback. i have added the control directly in module `llm_openai`.

---

Thanks, will check